include/miopen/miopen.h

```diff
@@ -8003,6 +8004,65 @@ MIOPEN_EXPORT miopenStatus_t miopenMultiMarginLossForward(miopenHandle_t handle,
 // CLOSEOUT LossFunction DOXYGEN GROUP
 #endif // MIOPEN_BETA_API
 
+#ifdef MIOPEN_BETA_API
+// Median APIs
+/** @addtogroup median
+ *
+ *  @{
+ */
+
+/*! @brief Execute a Median forward layer
+ *
+ * @param handle                   MIOpen handle (input)
+ * @param inputDesc                Tensor descriptor for input tensor (input)
+ * @param input                    Data tensor input (input)
+ * @param outputDesc               Tensor descriptor for output tensor (input)
+ * @param output                   Data tensor output (output)
+ * @param indicesDesc              Tensor descriptor for indices tensor (input)
+ * @param indices                  Data tensor indices (output)
+ * @param dim                      The dimension to reduce (input)
+ * @param keepdim                  Whether the output tensor has dim
+                                   retained or not (Default=False) (input)
+ * @return                         miopenStatus_t
+ */
+MIOPEN_EXPORT miopenStatus_t miopenMedianForward(miopenHandle_t handle,
+                                                 const miopenTensorDescriptor_t inputDesc,
+                                                 const void* input,
+                                                 const miopenTensorDescriptor_t outputDesc,
+                                                 void* output,
+                                                 const miopenTensorDescriptor_t indicesDesc,
+                                                 size_t* indices,
```

Comment:
```suggestion
                                                 uint64_t* indices,
```

Reply:
It would be really helpful if you could provide a clear justification for your recommendation. At the moment, I'm uncertain about when to use `size_t` vs. `uint64_t`. I understand that `uint64_t` guarantees 64-bit params across all machines. However, semantically, `size_t` seems more appropriate as it represents a type large enough to hold the size of any possible object. Currently, I don’t see consistent usage of `size_t` and `uint64_t` in MIOpen, which leads me to assume that these two types are used interchangeably.

Reply:
My point is keeping the consistentcy. You will need to have a datatype check for indicesDesc, and of course, check whether its datatype is `miopenInt64` (I don't clearly remember that enum value name).

Reply:
On a second thought, `indices` should support other int datatypes not just `uint64_t`. Because people may have needed for a let bit datatype for better performance. Of course you will need to check is that datatype suitable (i.e. `uint8_t` works `inputDesc.GetSizes()[dim] < 256`)

Reply:
I know that supporting multiple data types for indices would indeed allow users to choose the type of indices they prefer, potentially improving performance. For example, using `uint32_t` instead of the default `uint64_t` could save half the memory required for storage and (potentially) enhance overall processing efficiency.

However, let’s consider how similar operations handle indices, such as [getitem](https://github.com/ROCm/MIOpen/blob/d172672a9baee41a7fcd9446ee631255a4737486/driver/getitem_driver.hpp#L53). In this case, the indices type is simply fixed to `int32_t`. While supporting multiple data types for indices is a valuable idea, I don’t believe it’s a strict requirement at this point. In my opinion, it would unnecessarily complicate the logic in MIOpen.

For context, MIOpen allows users to specify input types through conventions like `median`/`medianfp16`/`medianbfp16`. If we were to support multiple data types for `indices`, how would users specify the `indices`'s desired type? There doesn’t appear to be any existing official guidance on this. For instance, while the `Gather` operations support multiple data types for indices, the driver API implementation by @cognaiger9 fixes the type rather than allowing users to choose [[ref]](https://github.com/cognaiger9/MIOpen/blob/9d2201d89b21a2fed35ead9857a2dde7e49024f4/driver/dm_gather.cpp#L32). You also suggested me one op of @hieule88, which is the `UnsortedSegmentSum`, but I checked it out, and this op's implementation forget to write the logic of choosing type for `index` tensor [[ref]](https://github.com/moreh-dev/MIOpen/pull/63/files#r1897129056). It appears that, despite having the logic to handle multiple data types for indices, the current MIOpen code structure lacks a proper and seamless way to utilize it effectively. Therefore, I propose simplifying the implementation by fixing indices to a single data type.

Additionally, other merged operations, such as `kthvalues`, also fix the indices type without further comments or issues [[ref]](https://github.com/ROCm/MIOpen/blob/d172672a9baee41a7fcd9446ee631255a4737486/include/miopen/miopen.h#L7938). Unless there are explicit requirements for supporting multiple data types for `indices`, it seems reasonable to treat this as a design decision: either allow flexibility or stick to a single fixed type.

Regarding your point:

> You will need to have a datatype check for indicesDesc, and of course, check whether its datatype is miopenInt64 (I don’t clearly remember that enum value name).

When I decide to fix the data type for indices, there would no longer be a need to check if it matches miopenInt64, because I explicitly assign type of `miopenInt64` to my `indices` tensor:
```
if(SetTensorNd(indicesDesc, output_dims, miopenInt64) != miopenStatusSuccess)
    MIOPEN_THROW("Error parsing indices tensor.");
```
The fixed type would ensure we consistently use a specific integer type, simplifying the implementation.

Reply:
You are confused about how MIOpen is used. Your above references are those from MIOpenDriver. MIOpenDriver actually is just a tool for benchmarking and verifcation, and it has no use for computation. You don't have to worry about the convenstion of somethings like `median/medianfp16/medianbfp16`, or wether they are enough, because they are just instances of operations.

Users of MIOpen only have access to API writen in `miopen.h`, via `MIOpenLib.so`, or `libMIOpen.so` (I don't remember which one). Doxygen comments in `miopen.h` actually are documentations, and they can be auto generate to (this)[https://rocm.docs.amd.com/projects/MIOpen/en/latest/]. You are right about the less descrition about supported datatypes, but don't worry, it has already been a problem of MIOpen.

Although there are operations from Moreh in MIOpen, they may not perfect. You can ask Hieu about a huge bug of `Adam` made by senior Jungkeun. Because of that, I don't think that we have to stricky follow the previous implementations, especially those are oblivious that could be made better.

Another reason that I suggest this change is that there are more than one reviewer from AMD, and they may have different point of view. When being reviewed they may suggest this change, or not. It would take a lot of effort if people who maintains that PR is not you, while you can quickly update it. I'm sure that you can make it quick because you understand the code and don't have to rebenchmark for other datatypes excluding `uint64_t`, just have to verify the correctness.

Reply:
I understand now. I'll make the changes as you suggested - sorry for the confusion!